Using a shared set of thread pools across multiple connections

sduskis opened the pull request:
There are cases where users create multiple connections.  They either did it for (hopefully) legacy performance issues or this was done through a complex distributed job, such as a dataflow job.  Either way, each connection doesn't need its own thread pool; the thread pools are auto-expanding, so if more threads are needed, they will automatically be allocated.  Shared thread pools will be much more efficient in such cases.

sduskis commented:
I'm not sure what happened to the previous PR.  I think I messed things up a bit.  I removed any reference to user based thread pools, and all of the management is done by us.

mgarolera commented on bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java:
```diff
@@ -200,109 +194,106 @@ public void run() {
         !Strings.isNullOrEmpty(options.getProjectId()), PROJECT_ID_EMPTY_OR_NULL);
-    Preconditions.checkArgument(
-        !Strings.isNullOrEmpty(options.getZoneId()), ZONE_ID_EMPTY_OR_NULL);
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(options.getZoneId()), ZONE_ID_EMPTY_OR_NULL);
```
nit: Not sure why this changed here but I slightly prefer it the old way, matching the other checkts. Up to you.

sduskis commented:
I updated the PR. I removed the increment/decrement logic.  The threadpools will stay around until the JVM dies. Users can customize the threadpooling by creating their own BigtableSessionSharedThreadPools implementation.

mgarolera commented:
LGTM, asides from the setInstance we discussed.

sduskis commented:
I removed setInstance. Merging.
